tools/gen_sig_info.py

```diff
@@ -0,0 +1,285 @@
+#include <GL/gl.h>
+#include <GL/glut.h>
+#include <GL/glew.h>
+#include <AL/al.h>
+#include <AL/alc.h>
+#include <SDL/SDL.h>
+#include <SDL/SDL_mutex.h>
+#include <SDL/SDL_image.h>
+#include <SDL/SDL_mixer.h>
+#include <SDL/SDL_surface.h>
+#include <SDL/SDL_ttf.h>
+#include <SDL/SDL_gfxPrimitives.h>
+#include <SDL/SDL_rotozoom.h>
+#include <webgl/webgl1_ext.h>
+#include <webgl/webgl2_ext.h>
+#include <X11/Xlib.h>
+#include <X11/Xutil.h>
+#include <uuid/uuid.h>
+''' % os.path.basename(__file__)
+
+footer = '''\
+};
+
+int main(int argc, char* argv[]) {
+  return argc + (intptr_t)symbol_list;
+}
+'''
+
+wasi_symbols = {
+  'proc_exit',
+  'environ_sizes_get',
+  'environ_get',
+  'clock_time_get',
+  'clock_res_get',
+  'fd_write',
+  'fd_pwrite',
+  'fd_read',
+  'fd_pread',
+  'fd_close',
+  'fd_seek',
+  'fd_sync',
+  'fd_fdstat_get',
+  'args_get',
+  'args_sizes_get',
+}
+
+
+def ignore_symbol(s):
+  if s in {'SDL_GetKeyState'}:
+    return True
+  if s.startswith('emscripten_gl') or s.startswith('emscripten_alc'):
+    return True
+  if s.startswith('egl'):
+    return True
+  if s.startswith('gl') and any(s.endswith(x) for x in ('NV', 'EXT', 'WEBGL', 'ARB', 'ANGLE')):
+    return True
+  return False
+
+
+def create_c_file(filename, symbol_list):
+  source_lines = [header]
+  source_lines.append('\nvoid* symbol_list[] = {')
+  for s in symbol_list:
+    if s in wasi_symbols:
+      source_lines.append(f'  &__wasi_{s},')
+    else:
+      source_lines.append(f'  &{s},')
+  source_lines.append(footer)
+  utils.write_file(filename, '\n'.join(source_lines) + '\n')
+
+
+def valuetype_to_chr(t, t64):
+  if t == webassembly.Type.I32 and t64 == webassembly.Type.I64:
+    return 'p'
+  assert t == t64
+  return {
+    webassembly.Type.I32: 'i',
+    webassembly.Type.I64: 'j',
+    webassembly.Type.F32: 'f',
+    webassembly.Type.F64: 'd',
+  }[t]
+
+
+def functype_to_str(t, t64):
+  assert len(t.returns) == len(t64.returns)
+  assert len(t.params) == len(t64.params)
+  if t.returns:
+    assert len(t.returns) == 1
+    rtn = valuetype_to_chr(t.returns[0], t64.returns[0])
+  else:
+    rtn = 'v'
+  for p, p64 in zip(t.params, t64.params):
+    rtn += valuetype_to_chr(p, p64)
+  return rtn
+
+
+def write_sig_library(filename, sig_info):
+  lines = [
+      '/* Auto-generated by tools/gen_sig_info.py. DO NOT EDIT. */',
+      '',
+      'sigs = {']
+  for s, sig in sorted(sig_info.items()):
+    lines.append(f"  {s}__sig: '{sig}',")
+  lines.append('}')
+  lines.append('mergeInto(LibraryManager.library, sigs);')
+  utils.write_file(filename, '\n'.join(lines) + '\n')
+
+
+def update_sigs(sig_info):
+  print("updating __sig attributes ...")
+  for sym, sig in sig_info.items():
+    files = glob.glob('src/*.js') + glob.glob('src/**/*.js')
+    replacement = f"s/{sym}__sig: '.*',/{sym}__sig: '{sig}',/"
+    # This is really hacky, but also should only be needed in the transiion period
+    shared.check_call(['perl', '-pi', '-e', replacement] + files)
```

Comment:
Can we not use a regex in python for this? How will it run on windows?

Reply:
As the comment says this isn't really designed to be a long term solution, its just a hack to help me move signatures from manual-generation to auto-generation.  Once that is process is done we can delete this code.
I will add more comments.

It will work on any system that has perl installed :)   And will fail if perl is not installed, even on linux.

Reply:
Sounds good. Maybe I'm just surprised perl can run on windows :wink:

Reply:
I took the bait the re-implemented in python.  Much faster now too!